File: reviewinn-backend/routers/view_tracking.py
"""
View Tracking Router - Industry Standard Implementation
Provides endpoints for tracking views with rate limiting and fraud prevention
"""
from fastapi import APIRouter, Depends, HTTPException, Request
from sqlalchemy.orm import Session
from typing import Dict, Any, Optional
from database import get_db
from models.user import User
from services.view_tracking_service import ViewTrackingService
from core.auth_dependencies import AuthDependencies

from pydantic import BaseModel
import logging

logger = logging.getLogger(__name__)
router = APIRouter()

class ViewTrackingResponse(BaseModel):
    """Response model for view tracking operations"""
    tracked: bool
    reason: str
    view_count: int
    message: str

class ViewAnalyticsResponse(BaseModel):
    """Response model for view analytics"""
    review_id: int
    total_views: int
    unique_users: int
    unique_sessions: int
    analytics: Dict[str, Any]

@router.post("/reviews/{review_id}/view", response_model=ViewTrackingResponse)
async def track_review_view(
    review_id: int,
    request: Request,
    db: Session = Depends(get_db),
    current_user: Optional[User] = Depends(AuthDependencies.get_current_user_optional)
):
    """
    Track a review view with industry-standard rate limiting.
    
    Features:
    - Only authenticated users can increment view counts
    - 24-hour rate limiting per user per review
    - Fraud prevention and suspicious activity detection
    - Session-based tracking for analytics
    
    Args:
        review_id: The ID of the review being viewed
        request: FastAPI request object for IP and user agent
        current_user: Optional authenticated user
    
    Returns:
        ViewTrackingResponse with tracking status and updated count
    """
    try:
        # Initialize view tracking service
        view_service = ViewTrackingService(db)
        
        # Track the view
        result = await view_service.track_review_view(
            review_id=review_id,
            request=request,
            user=current_user
        )
        
        # Determine appropriate message based on result
        if result["tracked"]:
            message = "View tracked successfully"
            logger.info(f"View tracked for review {review_id} by user {current_user.user_id if current_user else 'anonymous'}")
        else:
            message = f"View not tracked: {result['reason']}"
            logger.debug(f"View not tracked for review {review_id}: {result['reason']}")
        
        return ViewTrackingResponse(
            tracked=result["tracked"],
            reason=result["reason"],
            view_count=result["view_count"],
            message=message
        )
        
    except Exception as e:
        logger.error(f"Error tracking view for review {review_id}: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail="Internal server error while tracking view"
        )

@router.post("/entities/{entity_id}/view", response_model=ViewTrackingResponse)
async def track_entity_view(
    entity_id: int,
    request: Request,
    db: Session = Depends(get_db),
    current_user: Optional[User] = Depends(AuthDependencies.get_current_user_optional)
):
    """
    Track an entity view with the same rate limiting as reviews.
    
    Args:
        entity_id: The ID of the entity being viewed
        request: FastAPI request object for IP and user agent
        current_user: Optional authenticated user
    
    Returns:
        ViewTrackingResponse with tracking status and updated count
    """
    try:
        # Initialize view tracking service
        view_service = ViewTrackingService(db)
        
        # Track the entity view
        result = await view_service.track_entity_view(
            entity_id=entity_id,
            request=request,
            user=current_user
        )
        
        # Determine appropriate message based on result
        if result["tracked"]:
            message = "Entity view tracked successfully"
            logger.info(f"Entity view tracked for entity {entity_id} by user {current_user.user_id if current_user else 'anonymous'}")
        else:
            message = f"Entity view not tracked: {result['reason']}"
            logger.debug(f"Entity view not tracked for entity {entity_id}: {result['reason']}")
        
        return ViewTrackingResponse(
            tracked=result["tracked"],
            reason=result["reason"],
            view_count=result["view_count"],
            message=message
        )
        
    except Exception as e:
        logger.error(f"Error tracking entity view for entity {entity_id}: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail="Internal server error while tracking entity view"
        )

@router.get("/reviews/{review_id}/analytics", response_model=ViewAnalyticsResponse)
async def get_review_analytics(
    review_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(AuthDependencies.get_current_user)
):
    """
    Get view analytics for a review.
    
    Note: This endpoint requires authentication and should be restricted
    to review authors, admins, or users with appropriate permissions.
    
    Args:
        review_id: The ID of the review
        current_user: Authenticated user
    
    Returns:
        ViewAnalyticsResponse with comprehensive analytics
    """
    try:
        # Initialize view tracking service
        view_service = ViewTrackingService(db)
        
        # Get analytics (this method should check permissions internally)
        analytics = await view_service.get_review_analytics(
            review_id=review_id,
            requesting_user=current_user
        )
        
        return ViewAnalyticsResponse(
            review_id=review_id,
            total_views=analytics["total_views"],
            unique_users=analytics["unique_users"],
            unique_sessions=analytics["unique_sessions"],
            analytics=analytics
        )
        
    except PermissionError as e:
        logger.warning(f"Permission denied for analytics access: {str(e)}")
        raise HTTPException(
            status_code=403,
            detail="You don't have permission to view these analytics"
        )
    except Exception as e:
        logger.error(f"Error getting analytics for review {review_id}: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail="Internal server error while retrieving analytics"
        )

@router.get("/health")
async def view_tracking_health():
    """Health check endpoint for view tracking service"""
    return {
        "status": "healthy",
        "service": "view_tracking",
        "version": "1.0.0",
        "features": [
            "24h_rate_limiting",
            "fraud_prevention", 
            "analytics_aggregation",
            "authenticated_only"
        ]
    }

File: reviewinn-frontend/src/api/viewTracking.ts
/**
 * View Tracking Service - Frontend Implementation
 * Handles view tracking with industry-standard rate limiting and best practices
 */
import { httpClient } from './httpClient';
import { useAuthStore } from '../stores/authStore';
import { API_CONFIG, API_ENDPOINTS } from './config';

export interface ViewTrackingResponse {
  tracked: boolean;
  reason: string;
  view_count: number;
  message: string;
}

export interface ViewAnalytics {
  review_id: number;
  total_views: number;
  unique_users: number;
  unique_sessions: number;
  analytics: {
    views_today: number;
    views_this_week: number;
    views_this_month: number;
    last_view_at: string | null;
    review_title?: string;
    review_rating?: number;
    entity_id?: number;
    created_at?: string;
  };
}

class ViewTrackingService {
  private viewedReviews = new Set<number>();
  private viewedEntities = new Set<number>();
  private lastViewTimes = new Map<string, number>();
  
  /**
   * Track a review view with industry-standard rate limiting
   * @param reviewId - The ID of the review being viewed
   * @returns Promise with tracking result
   */
  async trackReviewView(reviewId: number): Promise<ViewTrackingResponse> {
    try {
      // Get authentication state from store
      const authState = useAuthStore.getState();
      
      // Determine rate limiting based on authentication status
      let rateLimitMs: number;
      let viewKey: string;
      
      if (authState.isAuthenticated) {
        // Authenticated users: 24 hour rate limit
        rateLimitMs = 24 * 60 * 60 * 1000;
        viewKey = `review_${reviewId}_user`;
      } else {
        // Anonymous users: 1 hour rate limit
        rateLimitMs = 1 * 60 * 60 * 1000;
        viewKey = `review_${reviewId}_anon`;
      }

      // Client-side rate limiting check
      const lastViewTime = this.lastViewTimes.get(viewKey);
      const now = Date.now();

      if (lastViewTime && (now - lastViewTime) < rateLimitMs) {
        const timeLeft = rateLimitMs - (now - lastViewTime);
        const hoursLeft = Math.ceil(timeLeft / (60 * 60 * 1000));
        const minutesLeft = Math.ceil(timeLeft / (60 * 1000));
        
        const timeMessage = hoursLeft > 0 
          ? `${hoursLeft} hour${hoursLeft !== 1 ? 's' : ''}` 
          : `${minutesLeft} minute${minutesLeft !== 1 ? 's' : ''}`;
          
        return {
          tracked: false,
          reason: `Rate limited - ${timeMessage} remaining`,
          view_count: 0,
          message: `You can view this review again in ${timeMessage}`
        };
      }

      // Check if already viewed in this session (prevent duplicate calls)
      if (this.viewedReviews.has(reviewId)) {
        return {
          tracked: false,
          reason: 'Already viewed in this session',
          view_count: 0,
          message: 'View already tracked this session'
        };
      }

      // Immediately mark as viewed to prevent race conditions
      this.viewedReviews.add(reviewId);
      this.lastViewTimes.set(viewKey, now);

      // Make API call to track the view with error handling
      try {
        const response = await httpClient.post<ViewTrackingResponse>(
          `${API_CONFIG.BASE_URL}${API_ENDPOINTS.VIEW_TRACKING.TRACK_REVIEW(reviewId.toString())}`
        );

        // If successfully tracked, save to localStorage for persistence
        if (response.data && response.data.tracked) {
          // State already updated above to prevent race conditions
          this.saveViewTimeToStorage(viewKey, now);
        } else {
          // If not tracked by server, revert the state changes
          this.viewedReviews.delete(reviewId);
          this.lastViewTimes.delete(viewKey);
        }

        return response.data || {
          tracked: false,
          reason: 'Invalid response',
          view_count: 0,
          message: 'Invalid server response'
        };
      } catch (apiError) {
        // If API call fails, still track locally to prevent UI issues
        console.warn('View tracking API call failed, tracking locally:', apiError);
        
        // State already updated above, just save to localStorage
        this.saveViewTimeToStorage(viewKey, now);
        
        return {
          tracked: true, // Report as tracked to prevent retries
          reason: 'Tracked locally (API unavailable)',
          view_count: 0,
          message: 'View tracked locally due to server issues'
        };
      }

    } catch (error) {
      console.error('Error tracking review view:', error);
      return {
        tracked: false,
        reason: 'Network error',
        view_count: 0,
        message: 'Failed to track view due to network error'
      };
    }
  }

  /**
   * Track an entity view with the same rate limiting as reviews
   * @param entityId - The ID of the entity being viewed
   * @returns Promise with tracking result
   */
  async trackEntityView(entityId: number): Promise<ViewTrackingResponse> {
    try {
      // Get authentication state
      const authState = useAuthStore.getState();
      
      // Determine rate limiting based on authentication status
      let rateLimitMs: number;
      let viewKey: string;
      
      if (authState.isAuthenticated) {
        // Authenticated users: 24 hour rate limit
        rateLimitMs = 24 * 60 * 60 * 1000;
        viewKey = `entity_${entityId}_user`;
      } else {
        // Anonymous users: 1 hour rate limit
        rateLimitMs = 1 * 60 * 60 * 1000;
        viewKey = `entity_${entityId}_anon`;
      }

      // Client-side rate limiting check
      const lastViewTime = this.lastViewTimes.get(viewKey);
      const now = Date.now();

      if (lastViewTime && (now - lastViewTime) < rateLimitMs) {
        const timeLeft = rateLimitMs - (now - lastViewTime);
        const hoursLeft = Math.ceil(timeLeft / (60 * 60 * 1000));
        const minutesLeft = Math.ceil(timeLeft / (60 * 1000));
        
        const timeMessage = hoursLeft > 0 
          ? `${hoursLeft} hour${hoursLeft !== 1 ? 's' : ''}` 
          : `${minutesLeft} minute${minutesLeft !== 1 ? 's' : ''}`;
          
        return {
          tracked: false,
          reason: `Rate limited - ${timeMessage} remaining`,
          view_count: 0,
          message: `You can view this entity again in ${timeMessage}`
        };
      }

      // Check if already viewed in this session
      if (this.viewedEntities.has(entityId)) {
        return {
          tracked: false,
          reason: 'Already viewed in this session',
          view_count: 0,
          message: 'View already tracked this session'
        };
      }

      // Make API call to track the view with error handling
      try {
        const response = await httpClient.post<ViewTrackingResponse>(
          `${API_CONFIG.BASE_URL}${API_ENDPOINTS.VIEW_TRACKING.TRACK_ENTITY(entityId.toString())}`
        );

        // If successfully tracked, update local state
        if (response.data && response.data.tracked) {
          this.viewedEntities.add(entityId);
          this.lastViewTimes.set(viewKey, now);
          
          // Store in localStorage for persistence
          this.saveViewTimeToStorage(viewKey, now);
        }

        return response.data || {
          tracked: false,
          reason: 'Invalid response',
          view_count: 0,
          message: 'Invalid server response'
        };
      } catch (apiError) {
        // If API call fails, still track locally to prevent UI issues
        console.warn('Entity view tracking API call failed, tracking locally:', apiError);
        
        this.viewedEntities.add(entityId);
        this.lastViewTimes.set(viewKey, now);
        this.saveViewTimeToStorage(viewKey, now);
        
        return {
          tracked: true, // Report as tracked to prevent retries
          reason: 'Tracked locally (API unavailable)',
          view_count: 0,
          message: 'View tracked locally due to server issues'
        };
      }

    } catch (error) {
      console.error('Error tracking entity view:', error);
      return {
        tracked: false,
        reason: 'Network error',
        view_count: 0,
        message: 'Failed to track view due to network error'
      };
    }
  }

  /**
   * Get analytics for a review (requires appropriate permissions)
   * @param reviewId - The ID of the review
   * @returns Promise with analytics data
   */
  async getReviewAnalytics(reviewId: number): Promise<ViewAnalytics | null> {
    try {
      const response = await httpClient.get<ViewAnalytics>(
        `${API_CONFIG.BASE_URL}${API_ENDPOINTS.VIEW_TRACKING.REVIEW_ANALYTICS(reviewId.toString())}`
      );
      return response.data || null;
    } catch (error) {
      console.error('Error fetching review analytics:', error);
      return null;
    }
  }

  /**
   * Save view time to localStorage for persistence across sessions
   * @param viewKey - The key for the view
   * @param timestamp - The timestamp of the view
   */
  private saveViewTimeToStorage(viewKey: string, timestamp: number): void {
    try {
      const storedViews = JSON.parse(localStorage.getItem('review_view_times') || '{}');
      storedViews[viewKey] = timestamp;
      localStorage.setItem('review_view_times', JSON.stringify(storedViews));
    } catch (error) {
      console.warn('Failed to save view time to localStorage:', error);
    }
  }

  /**
   * Load view times from localStorage on initialization
   */
  private loadViewTimesFromStorage(): void {
    try {
      const storedViews = JSON.parse(localStorage.getItem('review_view_times') || '{}');
      const now = Date.now();
      const rateLimitMs = 24 * 60 * 60 * 1000; // 24 hours

      // Only load view times that are still within the rate limit window
      Object.entries(storedViews).forEach(([key, timestamp]) => {
        if (typeof timestamp === 'number' && (now - timestamp) < rateLimitMs) {
          this.lastViewTimes.set(key, timestamp);
        }
      });

      // Clean up old entries from localStorage
      const validViews: Record<string, number> = {};
      this.lastViewTimes.forEach((timestamp, key) => {
        validViews[key] = timestamp;
      });
      localStorage.setItem('review_view_times', JSON.stringify(validViews));

    } catch (error) {
      console.warn('Failed to load view times from localStorage:', error);
    }
  }

  /**
   * Check if a review can be viewed (client-side check)
   * @param reviewId - The ID of the review
   * @returns Whether the review can be viewed
   */
  canViewReview(reviewId: number): boolean {
    const authState = useAuthStore.getState();
    
    // Determine rate limiting based on authentication status
    let rateLimitMs: number;
    let viewKey: string;
    
    if (authState.isAuthenticated) {
      rateLimitMs = 24 * 60 * 60 * 1000; // 24 hours for authenticated users
      viewKey = `review_${reviewId}_user`;
    } else {
      rateLimitMs = 1 * 60 * 60 * 1000; // 1 hour for anonymous users
      viewKey = `review_${reviewId}_anon`;
    }

    const lastViewTime = this.lastViewTimes.get(viewKey);
    if (!lastViewTime) {
      return true;
    }

    return (Date.now() - lastViewTime) >= rateLimitMs;
  }

  /**
   * Check if an entity can be viewed (client-side check)
   * @param entityId - The ID of the entity
   * @returns Whether the entity can be viewed
   */
  canViewEntity(entityId: number): boolean {
    const authState = useAuthStore.getState();
    
    // Determine rate limiting based on authentication status
    let rateLimitMs: number;
    let viewKey: string;
    
    if (authState.isAuthenticated) {
      rateLimitMs = 24 * 60 * 60 * 1000; // 24 hours for authenticated users
      viewKey = `entity_${entityId}_user`;
    } else {
      rateLimitMs = 1 * 60 * 60 * 1000; // 1 hour for anonymous users
      viewKey = `entity_${entityId}_anon`;
    }

    const lastViewTime = this.lastViewTimes.get(viewKey);
    if (!lastViewTime) {
      return true;
    }

    return (Date.now() - lastViewTime) >= rateLimitMs;
  }

  /**
   * Clear all view tracking data (useful for logout)
   */
  clearViewData(): void {
    this.viewedReviews.clear();
    this.viewedEntities.clear();
    this.lastViewTimes.clear();
    try {
      localStorage.removeItem('review_view_times');
    } catch (error) {
      console.warn('Failed to clear view data from localStorage:', error);
    }
  }

  /**
   * Initialize the service (load persisted data)
   */
  initialize(): void {
    this.loadViewTimesFromStorage();
  }
}

// Create and initialize the service instance
export const viewTrackingService = new ViewTrackingService();

// Initialize on module load
viewTrackingService.initialize();

export default viewTrackingService;
